syntax = "proto3";

package transport_catalogue_serialize;

message Coordinates {
	double lat = 1;
	double lng = 2;
};

message AvailableRoutes {
	uint32 id_form = 1;
	repeated string route = 2;
};

message Stop {
	uint32 id = 1;
	string name = 2;
	Coordinates coordinates = 3;
	AvailableRoutes aviable_routes = 4;
};

message StopsList {
	repeated Stop stops = 1;
};	

message Bus {
	string route_number = 1;
	bool is_circular = 2;
	repeated uint32 stop_id = 3;
};

message BusesList {
	repeated Bus bus = 1;
};	

message StopsDistance {
	uint32 id_form = 1;
	uint32 id_to = 2;
	uint32 distance = 3;
};

message DistancesList {
	repeated StopsDistance distances = 1;
};

message RoutingSettings {
	uint32 bus_velocity = 1;
	uint32 bus_wait_time = 2;
};

message RGB_A {
	uint32 red = 1;
	uint32 green = 2;
	uint32 blue = 3;
	bool is_rgba = 4;
	double opacity = 5;
};

message Color {
	bool monostate = 1;
	RGB_A representation_rgb_a = 2;
	string representation_string = 3;
};

message RenderSettings {
	double width = 1;
	double height = 2;
	double padding = 3;
	double line_width = 4;
	double stop_radius = 5;
	uint32 bus_label_font_size = 6;
	repeated double bus_label_offset = 7;
	uint32 stop_label_font_size = 8;
	repeated double stop_label_offset = 9;
	Color underlayer_color = 10;
	double underlayer_width = 11;
	repeated Color color_palette = 12;
};

message CatalogData {
	StopsList stops_list = 1;
	BusesList buses_list = 2;
	DistancesList distances_list = 3;
};

message TransportCatalogue {
	CatalogData catalog_data = 1;
	RoutingSettings routing_settings = 2;
	RenderSettings render_settings = 3;
};
